iml-agent/src/action_plugins/stratagem/server.rs

```diff
@@ -26,71 +28,153 @@ impl StratagemGroup {
     }
 }
 
+/// A rule to match over.
 #[derive(Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
 pub struct StratagemRule {
     pub action: String,
     pub expression: String,
     pub argument: String,
 }
 
+/// The top-level config
 #[derive(Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
-pub struct StratagemData {
+pub struct StratagemConfig {
     pub dump_flist: bool,
     pub groups: Vec<StratagemGroup>,
     pub device: StratagemDevice,
 }
 
-impl StratagemData {
+impl StratagemConfig {
     pub fn get_group_by_name(&self, name: &str) -> Option<&StratagemGroup> {
         self.groups.iter().find(|g| g.name == name)
     }
 }
 
+/// Contains matching results.
 #[derive(Debug, serde::Serialize, serde::Deserialize)]
 pub struct StratagemCounter {
     pub name: String,
     pub count: u64,
     pub have_flist: bool,
-    #[serde(flatten)]
-    pub extra: HashMap<String, serde_json::Value>,
 }
 
+/// A result for a `LAT_ATTR_CLASSIFY` rule.
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct StratagemClassifyResult {
+    pub attr_type: String,
+    pub have_flist: bool,
+    pub counters: Vec<StratagemCounter>,
+}
+
+/// A nested Counter used for matches to `LAT_ATTR_CLASSIFY`.
+/// This is nested within `StratagemClassifyResult` and considerably
+/// complicates the type hierarchy.
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct StratagemClassifyCounter {
+    pub name: String,
+    pub count: u64,
+    pub have_flist: bool,
+    pub expression: String,
+    pub classify: StratagemClassifyResult,
+}
+
+/// A result for a given rule group.
 #[derive(Debug, serde::Serialize, serde::Deserialize)]
 pub struct StratagemGroupResult {
     pub name: String,
-    pub counters: Vec<StratagemCounter>,
+    pub counters: Vec<StratagemCounters>,
+}
+
+/// Possible counter results.
+/// `StratagemClassifyCounter` matches `LAT_ATTR_CLASSIFY`.
+/// `StratagemCounter` matches everything else.
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+#[serde(untagged)]
+pub enum StratagemCounters {
+    StratagemClassifyCounter(StratagemClassifyCounter),
+    StratagemCounter(StratagemCounter),
 }
 
 #[derive(Debug, serde::Serialize, serde::Deserialize)]
 pub struct StratagemResult {
     pub group_counters: Vec<StratagemGroupResult>,
 }
 
-trait HaveFlist {
+/// Abstracts over the fact that `StratagemClassifyCounter`
+/// and `StratagemCounter` are mostly the same.
+///
+/// Exposes their common properties as trait methods.
+pub trait Counter {
+    fn name(&self) -> &str;
+    fn count(&self) -> u64;
     fn have_flist(&self) -> bool;
 }
 
-impl HaveFlist for StratagemCounter {
+impl Counter for &StratagemCounter {
```

Comment:
One thing we could consider doing here is using a macro (perhaps this is overkill but i'll mention it any way). As an example:

```
macro_rules! impl_Counter {
    ($T:ident) => {
        impl Counter for $T {
            fn name(&self) -> &str {
                &self.name
            }

            fn count(&self) -> u64 {
                self.count
            }
        }
    }
}
```

Then call the macro to define the implementation for each counter type:

```
impl_Counter!(&StratagemCounter)
impl_Counter!(&StratagemClassifyCounter)
```

I thought this might be a way to dry it up a bit but maybe it's overkill. What are your thoughts?

Reply:
I'd prefer to stay away from implementing our own macros unless there is a really clear benefit to doing so. I think they have a number of downsides.

Reply:
ok, i'll keep that in mind.